source/windows/secure_channel_tls_handler.c

```diff
@@ -939,73 +935,105 @@ static int s_do_client_side_negotiation_step_2(struct aws_channel_handler *handl
+                    sc_handler->read_extra = input_buffers[3].cbBuffer;
+                    AWS_LOGF_TRACE(
+                        AWS_LS_IO_TLS,
+                        "id=%p: Extra (incomplete) message received with length %llu.",
+                        (void *)handler,
+                        (unsigned long long)sc_handler->read_extra);
+                    has_extra = true;
+                } else {
+                    /* this means we processed everything in the buffer. */
+                    sc_handler->buffered_read_in_data_buf.len = 0;
+                    AWS_LOGF_TRACE(
+                        AWS_LS_IO_TLS,
+                        "id=%p: Decrypt ended exactly on the end of the record, resetting buffer.",
+                        (void *)handler);
+                }
             }
-
-            assert(
-                decrypted_length <=
-                sc_handler->buffered_read_out_data_buf.capacity - sc_handler->buffered_read_out_data_buf.len);
-            memcpy(
-                sc_handler->buffered_read_out_data_buf.buffer + sc_handler->buffered_read_out_data_buf.len,
-                (uint8_t *)input_buffers[1].pvBuffer,
-                decrypted_length);
-            sc_handler->buffered_read_out_data_buf.len += decrypted_length;
         }
-
-        return AWS_OP_SUCCESS;
         /* SEC_E_INCOMPLETE_MESSAGE means the message we tried to decrypt isn't a full record and we need to
            append our next read to it and try again. */
-    } else if (status == SEC_E_INCOMPLETE_MESSAGE) {
-        sc_handler->estimated_incomplete_size = input_buffers[1].cbBuffer;
-        AWS_LOGF_TRACE(
-            AWS_LS_IO_TLS,
-            "id=%p: (incomplete) message received. Expecting remaining portion of size %llu.",
-            (void *)handler,
-            (unsigned long long)sc_handler->estimated_incomplete_size);
-        return aws_raise_error(AWS_IO_READ_WOULD_BLOCK);
-    } else {
-        AWS_LOGF_ERROR(
-            AWS_LS_IO_TLS, "id=%p: Error decypting message. SECURITY_STATUS is %d.", (void *)handler, (int)status);
-        int aws_error = s_determine_sspi_error(status);
-        aws_raise_error(aws_error);
-        return AWS_OP_ERR;
-    }
+        else if (status == SEC_E_INCOMPLETE_MESSAGE) {
+            sc_handler->estimated_incomplete_size = input_buffers[1].cbBuffer;
+            AWS_LOGF_TRACE(
+                AWS_LS_IO_TLS,
+                "id=%p: (incomplete) message received. Expecting remaining portion of size %llu.",
+                (void *)handler,
+                (unsigned long long)sc_handler->estimated_incomplete_size);
+            memmove(
+                sc_handler->buffered_read_in_data_buf.buffer,
+                sc_handler->buffered_read_in_data_buf.buffer + offset,
+                read_len);
+            sc_handler->buffered_read_in_data_buf.len = read_len;
+            aws_raise_error(AWS_IO_READ_WOULD_BLOCK);
+            error = AWS_OP_ERR;

```

Comment:
Do we need to worry about the do-while loop checking `error`?

Reply:
discussed offline, read extra is the only thing we care about here.